### PR TITLE
Handle errors in the theme filter fetch

### PR DESCRIPTION
**Pull request by noahtallen**
### Proposed Changes
This _might_ fix the timeout issues on the /themes route! After merging #69538, I dug into the data and found out that the timeout _always_ happens in the fetchThemeFilters middleware.

After I started looking at this code more carefully, I realized that it only ever calls `next` after the filters exist in the store:

```js
const unsubscribe = store.subscribe( () => {
	if ( Object.keys( getThemeFilters( store.getState() ) ).length > 0 ) {
		unsubscribe();
		return next();
	}
} );
```

So if the filters will never exist, this loop will run forever and the middleware will never continue. If some kind of network error happens, the theme filters will never exist -- in other words, we need error handling.

This PR adds that by looking for the "fetch theme filters" notice that gets created by the error handler for the theme filter fetch. 

We then create an error and pass it to `next`. This triggers the error handling behavior in our SSR middleware chain, which effectively means:
1. We get automatically logging of the error
2. Future midleware are skipped
3. The page is still rendered, but not with an SSR layout. The client will have to fetch themes and theme filters.

### Testing Instructions
You can replicate this issue locally on trunk like so:

1. In your hosts file, write `0.0.0.0 public-api.wordpress.com` and comment out any other entries which sandbox the public-api. This will effectively make every request to public-api fail.
4. Run `ping public-api.wordpress.com`. You should see a socket connection error if the above step worked.
5. Run `DEBUG=calypso:themes,calypso:pages,calypso:isomorphic-routing yarn start`
6. In an incognito/private browser window, open `calypso.localhost:3000/themes`
7. The page **should never load**. Wait for like 30 seconds and close the tab  -- verify that you see a performanceMark object logged to the console with `fetchThemeFilters` being the biggest culprit:

image

8. Stop the server, checkout this branch, and run `DEBUG=calypso:themes,calypso:pages,calypso:isomorphic-routing yarn start`
9. In an incognito/private browser window, open `calypso.localhost:3000/themes`
10. This time, it might still take maybe 10 seconds for the page to load, but it will load. You should see a message about an error fetching theme filters in the console:

image

The page should load.

Also note the error handling. You'll see a message like this, which will ultimately show up in logstash once deployed. I encoded the error with JSON.stringify since we can't be completely sure of its type. (Other than that it's not a JS Error object)

```sh
23:08:53.174Z  WARN calypso: Theme fetch failed: {"errno":-61,"code":"ECONNREFUSED","syscall":"connect","address":"0.0.0.0","port":443} (reqId=fb8f4e37-1fbf-4f29-9fc7-543059d87e6f, url=/themes/, env=development, userAgent="Firefox 107", path=/themes/)
    Error: Theme fetch failed: {"errno":-61,"code":"ECONNREFUSED","syscall":"connect","address":"0.0.0.0","port":443}
        at /Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/my-sites/themes/controller.jsx:111:17
        at dispatch (/Users/noahallen/source/wp-calypso/node_modules/redux/lib/redux.js:305:7)
        at next (/Users/noahallen/source/wp-calypso/node_modules/redux-dynamic-middlewares/lib/index.js:17:83)
        at /Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom-api-middleware.js:63:11
        at /Users/noahallen/source/wp-calypso/node_modules/redux-thunk/lib/index.js:14:16
        at map (/Users/noahallen/source/wp-calypso/node_modules/redux/lib/redux.js:667:28)
        at Array.map (<anonymous>)
        at next (/Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom-http/utils.js:190:25)
        at themeFiltersHandlers (/Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom-http/utils.js:116:10)
        at /Users/noahallen/source/wp-calypso/build/webpack:/wp-calypso/client/state/data-layer/wpcom/theme-filters/index.js:45:11
```

### CODE REVIEW

**Comment by noahtallen** [comment]
The search bar seems to be messed up in dev mode:

### Dev mode:
image

### Production:
image

This is also a problem on trunk, so it shouldn't be related to this change.

**Comment by noahtallen** [comment]
cc @jsnajdr @tyxla, would love a review on this one. I would like to merge ASAP (to verify that fix in prod), and then handle follow-ups afterwards. (For example, it'd be nice to capture the network error, but that would be a bit of extra work)

**Comment by tyxla** [comment]
Sure thing @noahtallen, that's on my review queue for today.

File: client/my-sites/themes/controller.jsx [hunk 1]
```diff
@@ -96,6 +97,12 @@ export function fetchThemeFilters( context, next ) {
 	}
 
 	const unsubscribe = store.subscribe( () => {
+		// A notice is added to the state as the "on error" handler in the getThemeFilters action.
+		if ( getNotice( store.getState(), 'fetch-theme-filters-error' ) ) {
```
Comment (tyxla) [inline client/my-sites/themes/controller.jsx]: It's a bummer and feels a bit fragile we have to peek into the notices to see if there was an error. I wonder if this is worth introducing another reducer subtree to the `themes` redux tree, called `themeFiltersError`, and it could be as simple as a boolean flag, then we would dispatch an action to make it `true` right where we call `reportRerror()` in the data layer. Sounds like it would be better and more reliable than betting on the notices state. That way we also won't have to create a new selector to fetch a particular notice. What do you think?

File: client/state/notices/test/selectors.js [hunk 2]
```diff
@@ -20,4 +20,44 @@ describe( 'selectors', () => {
 			expect( notices ).toEqual( [ { noticeId: 1 }, { noticeId: 2 }, { noticeId: 3 } ] );
 		} );
 	} );
+
+	describe( 'getNotice()', () => {
```
Comment (tyxla) [inline client/state/notices/test/selectors.js]: Thanks for adding tests! Looks like these are wonderful candidates for a separate PR.

File: client/state/notices/selectors.js [hunk 3]
```diff
@@ -25,3 +25,8 @@ export const getNoticeLastTimeShown = createSelector(
 	( state, noticeId ) => state.notices.lastTimeShown[ noticeId ] || 0,
 	( state ) => state.notices.lastTimeShown
 );
+
+export const getNotice = createSelector(
```
Comment (tyxla) [inline client/state/notices/selectors.js]: Should this be part of a separate PR?

**Comment by tyxla** [comment]
For what it's worth, this tested well for me and I reproduced the test instructions perfectly. I just wonder if we can make it more reliable and explicit by having a dedicated theme filter error flag, rather than relying on notices.

File: client/my-sites/themes/controller.jsx (hunk 1 above)
Comment (noahtallen) [reply client/my-sites/themes/controller.jsx]: I was also thinking about this. I think it would be useful to save the error message as well, so we can report it in the server logs.

That said, I don't think it's technically any more or less fragile -- the on error handler for the request currently dispatches an action which happens to set the notice. This different approach would still dispatch an action, it'd just set a different part of the state.

But I'll try it anyways, since it at least makes more sense!

File: client/state/notices/selectors.js (hunk 3 above)
Comment (noahtallen) [reply client/state/notices/selectors.js]: The change in this PR relies on `getNotice`, so no. But if we use a different approach, then we can delete it altogether!

**Comment by noahtallen** [comment]
@tyxla, I've implemented your approach with a new `THEME_FILTERS_REQUEST_FAILURE` action, which is dispatched on request failure and stores the error in a new reducer. I also added a selector to read that value. Finally, in the controller file, we use that selector in the store subscriber to get the error.

The exact same testing steps should work. Next, I'll work to log the given error so that we can use it.

**Comment by noahtallen** [comment]
I've also added error handling, so these errors will show up in logstash now. I think this is ready to go!

Also, the ETK change is probably just because we're modifying the themes store, which is probably widely used in Calypso, and so imported somehow by ETK.

**Review by tyxla — APPROVED** [review]
This looks really good and works well in my testing 👍 

Let's 🚀 and keep an eye on the logs!
